logcheck/pkg/logcheck.go

```diff
@@ -496,3 +508,37 @@ func checkForIfEnabled(i *ast.IfStmt, pass *analysis.Pass, c *config) {
 			funcCall, varName, funcCall, varName, varName),
 	})
 }
+
+func checkForVerboseZero(expr *ast.CallExpr, pass *analysis.Pass) {
+	if typeAndValue, ok := pass.TypesInfo.Types[expr]; ok {
+		switch t := typeAndValue.Type.(type) {
+		case *types.Named:
+			if typeName := t.Obj(); typeName != nil {
+				if pkg := typeName.Pkg(); pkg != nil {
+					if typeName.Name() == "Verbose" && pkg.Path() == "k8s.io/klog/v2" {
+						lit, ok := expr.Args[0].(*ast.BasicLit)
+						if !ok || lit.Value != "0" {
+							return
+						}
+						msg := fmt.Sprintf("klog does not allow verbose zero")
+						pass.Report(analysis.Diagnostic{
+							Pos:     expr.Fun.Pos(),
+							Message: msg,
+						})
+					}
+					if typeName.Name() == "Logger" && pkg.Path() == "github.com/go-logr/logr" {
+						lit, ok := expr.Args[0].(*ast.BasicLit)
+						if !ok || lit.Value != "0" {
```

Comment:
This seems to work for `V(0)`, but only for a literal 0. Is it possible to check for the value?

Constants of value zero are currently not detected. I am undecided whether they should be, though: one the one hand, they cause the same overhead. On the other hand, the code using the constant was perhaps written that way with the intention to change the parameter value back and forth, in which case the tool should allow it without warning. What do you think?

Anyway, here's the test patch that I used to check:

```diff
diff --git a/logcheck/testdata/src/doNotAllowVerboseZeroLogs/doNotAllowVerboseZeroLogs.go b/logcheck/testdata/src/doNotAllowVerboseZeroLogs/doNotAllowVerboseZeroLogs.go
index 9b5855e..fee270e 100644
--- a/logcheck/testdata/src/doNotAllowVerboseZeroLogs/doNotAllowVerboseZeroLogs.go
+++ b/logcheck/testdata/src/doNotAllowVerboseZeroLogs/doNotAllowVerboseZeroLogs.go
@@ -28,16 +28,30 @@ import (
 
 var l, logger logr.Logger
 
+const (
+	zeroConst = 0
+	oneConst  = 1
+)
+
+var (
+	zeroVar klog.Level = 0
+	oneVar  klog.Level = 1
+)
+
 func verboseLogging() {
 
-	klog.V(0).Info("test log")                 // want `klog does not allow verbose zero`
-	klog.V(0).Infof("test log")                // want `klog does not allow verbose zero`
-	klog.V(0).Infoln("test log")               // want `klog does not allow verbose zero`
-	klog.V(0).InfoS("I'm logging at level 0.") // want `klog does not allow verbose zero`
+	klog.V(0).Info("test log")                         // want `klog does not allow verbose zero`
+	klog.V(0).Infof("test log")                        // want `klog does not allow verbose zero`
+	klog.V(0).Infoln("test log")                       // want `klog does not allow verbose zero`
+	klog.V(0).InfoS("I'm logging at level 0.")         // want `klog does not allow verbose zero`
+	klog.V(zeroConst).InfoS("I'm logging at level 0.") // want `klog does not allow verbose zero`
+	klog.V(zeroVar).InfoS("I'm logging at level 0.")
 	klog.V(1).Info("test log")
 	klog.V(1).Infof("test log")
 	klog.V(1).Infoln("test log")
 	klog.V(1).InfoS("I'm logging at level 1.")
+	klog.V(oneConst).InfoS("I'm logging at level 1.")
+	klog.V(oneVar).InfoS("I'm logging at level 1.")
 	klog.Info("test log")
 	klog.Infof("test log")
 	klog.Infoln("test log")
```

Reply:
I  think that constant zero needs to be checked, variables are not checked